srml/balances/src/tests.rs

```diff
@@ -110,7 +110,7 @@ fn lock_reasons_should_work() {
 		assert_ok!(<Balances as TransferAsset<_>>::withdraw(&1, 1, WithdrawReason::TransactionPayment));
 
 		Balances::set_lock(ID_1, &1, 10, u64::max_value(), WithdrawReason::TransactionPayment.into());
-		assert_ok!(<Balances as TransferAsset<_>>::transfer(&1, &2, 1));
+		assert_ok!(Balances::ensure_account_can_withdraw(&1, 1, WithdrawReason::Transfer, 9));
```

Comment:
These are not logically equivalent. Any such changes should be well-documented in the PR with why unit tests are being altered.

Reply:
The reason of the change is old behavior is not correct.
balances transfer method not only withdraw for transfer reason, but also withdraw for tx fee reason. Therefore it should fail, which indeed failed with the updated behavior.
Related: #1962, #1978

Reply:
@shaopengw maybe should have `assert_err!(<Balances as TransferAsset<_>>::transfer(&1, &2, 1));`

Reply:
@xlc The `assert_noop!` below has covered the `TransactionPayment` withdraw should fail case, so it should be fine without adding assert err for `transfer`.

If these tests are mean to be unit tests, maybe other `transfer` calls in this test should also be changed to `ensure_account_can_withdraw` like this line, as `transfer` implies both `WithdrawReason::TransactionPayment` and `WithdrawReason::Transfer`.

Reply:
> The reason of the change is old behavior is not correct.
balances transfer method not only withdraw for transfer reason, but also withdraw for tx fee reason

The old behaviour *is* correct. The only time transaction payment happens is prior to extrinsic execution in `executive`. Any further withdrawals are not "transaction payment" but high-level actions that are logically unrelated. The two should not be confused.